pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,140 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Package contains all the metadata of the package and drives package searching and browsing,
+// and support the fork/rebase upgrade workflow. It can be used by something like an app registry.
+type Package struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Name of an instance of the package.
+	// There may be multiple instances with the same package and different packageVersions.
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Description of the package.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// An pointer to the icon.
+	Icon string `json:"icon,omitempty" yaml:"icon,omitempty"`
+
+	// Search keywords for the package.
+	Keywords []string `json:"keywords,omitempty" yaml:"keywords,omitempty"`
+
+	// Homepage of the application package.
+	Home string `json:"home,omitempty" yaml:"home,omitempty"`
+
+	// Source specifies the upstream URL, e.g. https://github.com/foo/bar.git, file://host/path, etc.
+	// hosting the resource files specified in Base and Overlays.
+	// This is useful in the fork/rebase workflow.
+	Source string `json:"source,omitempty" yaml:"source,omitempty"`
+
+	// Version of the package.
+	PackageVersion string `json:"packageVersion,omitempty" yaml:"packageVersion,omitempty"`
+}
+
+// Manifest has all the information to expand of generate the k8s api resources.
+// It can be used by kubectl or some other tooling.
+// A manifest could be either a Base or an Overlay.
+type Manifest struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Name of an instance of the manifest.
+	// There may be multiple instances with the same manifest and different manifestVersions.
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Description of the manifest.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// Version of the manifest.
+	ManifestVersion string `json:"manifestVersion,omitempty" yaml:"manifestVersion,omitempty"`
```

Comment:
I'm not sure ManifestVersion is needed. If in doubt, we should leave it out. How do you imagine it would be used?

Imagine a package bundle available via git or a registry.

It would have both a Package instance and a Manifest instance for the bases. If the package maintainer updated the package, I'd expect them to update the packageVersion in Package regardless what else changed, whether it's the Package, Manifest, some K8s resource referenced by the Manifest, raw ConfigMap data, or anything else that would cause a change in behavior or format.

(I wouldn't expect overlays in such a package. Those would generally be created by the consumer of the package.)

Reply:
I added a version for manifest because I saw https://github.com/kubernetes/kubernetes/pull/52570#discussion_r140313676.
IMO it may be helpful for users to maintain their overlay. I imagine it is an Overlay only field?

Reply:
I wouldn't expect overlays to be present in an off-the-shelf package.

So this doesn't answer the question of how it would be used. Would it be added as a label to all resources in the package variant/instance? That would make isolated resource changes pretty disruptive in the case that the version was bumped.

I could imagine some users wanting some kind of opinionated versioning model, but I don't think we know what it would be yet, so I'd leave this out.

How about add a TODO here to figure this out, and add it to Jeff's task list?